app/src/main/java/com/example/androidtrainingproject/ui/home/HomeViewModel.kt

```diff
@@ -51,11 +57,52 @@ class HomeViewModel @Inject constructor(
         val newFilteredProducts = allProducts.filter { product: ProductResponse ->
             product.title.contains(newValue, ignoreCase = true)
         }
-        filteredProducts = newFilteredProducts
+        filteredProducts.clear()
+        filteredProducts.addAll(newFilteredProducts)
     }
 
     fun clearSearchText() {
         _searchText.value = ""
-        filteredProducts = allProducts
+        filteredProducts.clear()
+        filteredProducts.addAll(allProducts)
+    }
+
+    fun applyFilters(
+        checkedCategories: MutableList<String>,
+        rating: Int,
+        price: ClosedFloatingPointRange<Float>
+    ) {
+        filteredProducts.clear()
+        filteredProducts.addAll(allProducts.filter { product ->
+            checkedCategories.contains(product.category)
+                && product.rating.toInt() >= rating
+                && product.price >= price.start && product.price <= price.endInclusive
+        })
+
+        getActiveFiltersCount(checkedCategories, rating, price)
```

Comment:
Note for an alternative solution: This isn't wrong but you can make it more reactive just like in React. If `checkedCategories`, `rating` and `price` are all `StateFlow`s you can write a `combine` for all of them and place it in the `init` of the view model, just like you would do a `useEffect` on multiple states in React, and do the calculations there.